### ISSUE TITLE
ALIAS polish

### ISSUE THREAD

**Issue by Habbie** [PowerDNS/pdns#3838]
* [ ] investigate report of wildcard ALIAS leading to three identical A records in expansion
* [ ] finish #3808
* [ ] #3831 was caused by us punting a query to nproxy even if the client was TCP. Need to fix this.
* [ ] move ALIAS testing records from example.com to a smaller zone that we -do- check in `verify-dnssec-zone`

**Comment by goestreicher** [comment]
So while testing the LDAP backend with added support for ALIAS it turned out that not setting "recursor" in the configuration file or on the command line causes the test to fail without more information. Core just assumes that the ALIAS record doesn't exists.

**Comment by cmouse** [comment]
Maybe add:
 - When resolving ALIAS record during AXFR, the reason why it fails should be reported.

now it only says
```
[2017-04-25 18:36:35] Apr 25 18:36:35 Error resolving for ALIAS google-public-dns-a.google.com., aborting AXFR
```

**Comment by chasebolt** [comment]
I am running into the ALIAS tcp issue `Jul 13 19:55:01 dnsproxy.cc: Error sending reply with sendmsg (socket=21): Bad file descriptor`.

PowerDNS Authoritative Server 4.0.3

is there a way to resolve this or is the current strategy to not use ALIAS yet?

**Comment by mfowlewebs** [comment]
> actually support TCP queries (currently they lead to Feb 02 12:07:12 dnsproxy.cc: Error sending reply with sendmsg (socket=19): Bad file descriptor)

This doesn't seem like it's documented or mentioned anywhere. 

> is there a way to resolve this or is the current strategy to not use ALIAS yet?

Yeah, this. What the hell? We had no idea our customers DNS resolving just wasn't working for them. It's hard to justify ALIAS as usable in these conditions.

Should we open a new ticket to track this specific issue? This doesn't seem like polish- this seems like base functionality, which ATM seems very broken.

**Comment by Habbie** [comment]
When we implement ANAME (hopefully soon), we'll do it in a different way that will suffer from none of these issues. I agree this is more than polish, but there are no plans to fix it for ALIAS. The best workaround is to AXFR very periodically as we do resolve correctly during AXFR.

**Comment by jallakim** [comment]
Just to have the history; there was a draft for ALIAS earlier this year (https://tools.ietf.org/html/draft-dnsop-eden-alias-rr-type-00) which didn't seem to get enough momentum. Shortly after a draft for ANAME followed (https://tools.ietf.org/html/draft-ietf-dnsop-aname-00), which seems to have gained more attention, and hopefully get's ratified as a standard.

**Comment by Habbie** [comment]
@jallakim yes, indeed - it did not get momentum because we asked the ALIAS author (Anthony) to join forces with us on ANAME instead, and he agreed. Thank you for documenting the context here :)

**Comment by jallakim** [comment]
Hah. How many beers did it take to make him join forces? #bribe

**Comment by Habbie** [comment]
> Hah. How many beers did it take to make him join forces? #bribe

I don't know, Evan and Anthony were at the same conference at the time, I wasn't there ;)

**Comment by felixsand** [comment]
So this is what triggered all those errors in our logs as soon as we activated ALIAS...
Still not possible to get this prioritized?

**Comment by digitalocean-engineering** [comment]
We definitely have interest in seeing support for ANAME records.  I'd be happy to help out with testing, if that will be of use.

**Comment by egon1024** [comment]
We definitely have interest in seeing support for ANAME records.  I'd be happy to help out with testing, if that will be of use.

**Comment by digitalocean-engineering** [comment]
This is definitely of interest - we'd like to be able to use this feature but cannot currently commit to architecting our DNS infrastructure to rely simply on AXFRs for propagating zone data.  A fix for the ALIAS record issue or support for ANAME records would be <very> useful.

**Comment by Nosxxx** [comment]
Any update?

**Comment by Habbie** [comment]
Update: we did fix TCP for ALIAS. ANAME work is pending.

**Comment by felixsand** [comment]
@Habbie great news! Is this available in a stable release already or in what release will it be available?

**Comment by Habbie** [comment]
#6337 is the backport to 4.1.x which will be released as part of auth 4.1.2 at some point. 

If you go to https://downloads.powerdns.com/autobuilt_browser/#/auth/ and find '4.1.1-8' (at some point builds higher than 8 might appear), this fix should be in it.

**Comment by Omeryl** [comment]
Hey!

I just got done speaking to @Habbie on the IRC channel (thanks for your time!)

I was wondering if it would be potentially possible for PowerDNS to internally resolve/flatten ALIAS records when it's a zone it is authoritative over.

I.e. the following:

```
c1.test.dev.		60	IN	TYPE65401 \# 16 05706161733104746573740364657600 ; ALIAS to paas1.test.dev
paas1.test.dev.		60	IN	A	1.1.1.1
paas2.test.dev.		60	IN	A	2.2.2.2
```

It'd be beneficial when testing with a zone that doesn't exist (as above). I can see a few problems related to efficiency if it had to go out -> hit Cloudflare -> hits us again. 

Thanks!

**Comment by tingvold** [comment]
@Habbie, how's the progress for implementing ANAME?

**Comment by Habbie** [comment]
The ANAME draft has expired in favour of other IETF dnsop proposals (such as SVCB/HTTPSSVC), and because of that we currently have no plan to work on ANAME in PowerDNS. Instead, I suspect we will polish ALIAS a bit more this year.

Depending on your needs, the existing ALIAS code, or the existing [LUA functionality](https://doc.powerdns.com/authoritative/lua-records/index.html), can likely solve your problems. If you'd like some help, please find us on IRC or mailing list via https://www.powerdns.com/opensource.html

**Comment by Habbie** [comment]
The ANAME draft has expired in favour of other IETF dnsop proposals (such as SVCB/HTTPSSVC), and because of that we currently have no plan to work on ANAME in PowerDNS. Instead, I suspect we will polish ALIAS a bit more this year.

Depending on your needs, the existing ALIAS code, or the existing [LUA functionality](https://doc.powerdns.com/authoritative/lua-records/index.html), can likely solve your problems. If you'd like some help, please find us on IRC or mailing list via https://www.powerdns.com/opensource.html


> I was wondering if it would be potentially possible for PowerDNS to internally resolve/flatten ALIAS records when it's a zone it is authoritative over.

You can accomplish that today by pointing `resolver=` at a Recursor (or unbound or BIND or ..) on localhost that knows that `test.dev` lives in your auth that is also reachable without going out on a cable.

**Comment by edo1** [comment]
> somebody reported on IRC that wildcard ALIAS does not work at all on direct query, appears to be true. 9/oct/2019 fresh report came in for this not working on 4.2.0.

It doesn't work for me as well. PowerDNS version 4.1.6-3 from Debian Buster.

wildcard + A works
wildcard + CNAME works
explicit + ALIAS works
wildcard + ALIAS doesn't work.

**Comment by zeha** [comment]
Currently ALIAS uses `DP->completePacket` to handle all the resolving and also relying to the client. This kinda prevents reusing all the existing codepaths for signing etc. I guess the current design was chosen to reap some optimization benefits, but I don’t know what these are.

@Habbie could you please enlighten me on this?

**Comment by Habbie** [comment]
That choice was made so that the thread holding a DB connection can exit. It was also less typing. I want to change this so it works like LUA (for which the code is already there, too!) so everything just works, just haven't gotten around to it.

**Comment by aj-gh** [comment]
Noting it here (as discussed): For a FQDN having an ALIAS record only the ALIAS “resolved target” is returned on an ANY query even though other types exist as well.
e.g.
```
foo.bar. IN ALIAS google.com.
foo.bar. IN TXT “test”
```
would only return the google.com resolved A/AAAA records on an "ANY" query, ignoring the (locally) also-present TXT record

**Comment by thibmac** [comment]
Hello, 

Were there any updates on the DNSSEC / ALIAS topic since it was last discussed in here? 

It was also debated [here](https://lists.dns-oarc.net/pipermail/dns-operations/2022-April/021636.html)

Thanks!

**Comment by pschichtel** [comment]
@Habbie I just hit the ALIAS limitation with wildcards. If I understand you correctly, LUA records would already behave correctly. Can I call a resolver from a LUA record?

**Comment by Habbie** [comment]
https://github.com/PowerDNS/pdns/wiki/Lua-Examples-(Authoritative)#using-lua-to-simulate-cname-like-behaviour-on-the-zone-apex

**Comment by joachimtingvold** [comment]
I attempted to implement the LUA apex setup, but I'm having a hard time getting it to work.

Setup consists of three servers;

- authdns-master (hidden master)
- authdns1
- authdns2

`enable-lua-records=yes` is set on all three servers. I can query for the `LUA` record on `authdns-master` just fine;

```
root@authdns-master:~# dig @localhost foobar.com a

; <<>> DiG 9.16.42-Debian <<>> @localhost foobar.com a
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 57260
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;foobar.com.     IN  A

;; ANSWER SECTION:
foobar.com.    3600  IN  A 10.10.10.10

;; Query time: 3 msec
;; SERVER: ::1#53(::1)
;; WHEN: Mon Sep 11 02:55:50 CEST 2023
;; MSG SIZE  rcvd: 58
```

If I enable debug on `authdns-master`, I see the following;

```
Sep 11 01:49:20 authdns-master pdns_server[178900]: Remote 2001:<snip> wants 'fooobar.com|A', do = 1, bufsize = 1232 (4096): packetcache MISS
Sep 11 01:49:20 authdns-master pdns_server[178900]: [stub-resolver] Doing stub resolving for 'web-cdn.foobar.net.|A', using resolvers: 2a02:<snip>::2, <snip>.0.2
Sep 11 01:49:20 authdns-master pdns_server[178900]: [stub-resolver] Question for 'web-cdn.foobar.net.|A' got answered by 2a02:<snip>::2
```

However, if I attempt to do this on `authdns1` or `authdns2`, I don't get 

```
root@authdns1:~# dig @localhost foobar.com a

; <<>> DiG 9.16.42-Debian <<>> @localhost foobar.com a
; (2 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 13561
;; flags: qr aa rd; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;foobar.com.     IN  A

;; AUTHORITY SECTION:
foobar.com.    1800  IN  SOA authdns1.foobar.com. hostmaster.foobar.com. 2023091116 1800 600 2419200 1800

;; Query time: 0 msec
;; SERVER: ::1#53(::1)
;; WHEN: Mon Sep 11 02:57:57 CEST 2023
;; MSG SIZE  rcvd: 111
```

If I enable logging on `authdns1`, I only see the following when doing two lookups in a row;

```
Sep 11 02:59:17 authdns1 pdns_server[181429]: Remote ::1 wants 'foobar.com|A', do = 0, bufsize = 1232 (4096): packetcache MISS
Sep 11 02:59:30 authdns1 pdns_server[181429]: Remote ::1 wants 'foobar.com|A', do = 0, bufsize = 1232 (4096): packetcache HIT
```

If I look in the database for `authdns1` and `authdns2` (PostgreSQL backend), I can see the `LUA` records just fine. I have double checked, and they are 100% identical to the records in the database of `authdns-master`.

```
powerdns-ns1=# select name,type,content from records where domain_id=1 and type='LUA';
    name    | type |                                                      content
------------+------+----------------------------------------------------------------------------------------------------------------------------------------
 foobar.com | LUA  | A ";local r=resolve('web-cdn.foobar.net', pdns.A) local t={} for _,v in ipairs(r) do table.insert(t, v:toString()) end return t"
 foobar.com | LUA  | AAAA ";local r=resolve('web-cdn.foobar.net', pdns.AAAA) local t={} for _,v in ipairs(r) do table.insert(t, v:toString()) end return t"
```

I initially thought it could be related to not having set the `resolver` configuration parameter. However, it seems like it picks the system resolvers if it's unset (which is also evident from the logs on `authdns-master`). I can also see during the startup on `authdns1` and `authdns2` that resolver is working fine;

```
Sep 11 02:31:35 authdns1 pdns_server[181355]: [stub-resolver] Doing stub resolving for 'auth-4.8.2.security-status.secpoll.powerdns.com.|TXT', using resolvers: 2a02:<snip>::2, <snip>.0.2
Sep 11 02:31:35 authdns1 pdns_server[181355]: [stub-resolver] Question for 'auth-4.8.2.security-status.secpoll.powerdns.com.|TXT' got answered by 2a02:<snip>::2
```

Do you have any ideas why this might not be working, @Habbie ?

**Comment by Habbie** [comment]
> Do you have any ideas why this might not be working, @Habbie ?

Please open a thread on https://github.com/PowerDNS/pdns/discussions/ and then remove this comment. Thanks!

**Comment by klaus-nicat** [comment]
> https://github.com/PowerDNS/pdns/wiki/Lua-Examples-(Authoritative)#using-lua-to-simulate-cname-like-behaviour-on-the-zone-apex

The LUA-ALIAS example was not restored after the vandalism. On purpose (not working anymore), or just forgotten? Then I would restore it (see https://github.com/PowerDNS/pdns/wiki/Lua-Examples-(Authoritative)/_history)

**Comment by Habbie** [comment]
It's in https://github.com/PowerDNS/pdns/wiki/Lua-Examples-(Authoritative-LUA-records) now

**Comment by klaus-nicat** [comment]
Is there any drawback using LUA (ie blocking of receiver-threads ...)?

**Comment by Habbie** [comment]
> Is there any drawback using LUA (ie blocking of receiver-threads ...)?

yes, it blocks a thread (a distributor thread, I believe)

**Comment by mkoskar** [comment]
Here to report that `ALIAS` RR TYPE on wildcard name doesn't work (no `A/AAAA` are synthesized) in `4.9.3` autoritative. Didn't want to create a separate bug issue for it since it is mentioned in the description of this one already. Please do so if you feel it should be dealt separately.

**Comment by haslersn** [comment]
@Habbie 

The [online signing documentation](https://doc.powerdns.com/authoritative/dnssec/modes-of-operation.html#online-signing) states

> Please note that the ALIAS record type is [not supported](https://doc.powerdns.com/authoritative/guides/alias.html#alias-and-dnssec) in this mode.

Is this still up to date? It seems this works just fine with PowerDNS 4.9.4. If I understood correctly, "online signing" is used when the zone is not presigned and has DNSSEC keys configured.

On the primary PowerDNS server, which uses online signing, aliases worked out of the box for me.

On the secondary PowerDNS server, it didn't work at first, but after running `pdnsutil unset-presigned` on the secondary zone and importing DNSSEC keys (which I exported from the primary), aliases seem to work just fine on the secondary server, too.

**Comment by Habbie** [comment]
> Is this still up to date?

Yes. It might work, but it won't be signed, which means it won't actually work when a resolver is validating (and your zone has a DS record in the parent, etc.)

**Comment by haslersn** [comment]
@Habbie You're right. The A/AAAA records (that follow an ALIAS) are not signed, both on the primary and secondary. Interestingly, these records were at first only served on the primary but not on the secondary. After I ran `pdnsutil unset-presigned` on the secondary zone, it started serving the records.

Since we obviously want signed records, I enabled `outgoing-axfr-expand-alias=yes` on the primary (and re-enabled PRESIGNED on the secondary zone). Now the records are signed **only** on the secondary but not in the primary. Is is possible to also fix this on the primary? It would be kind of strange if it can only work on the secondary.

**Comment by Habbie** [comment]
What you are experiencing is the exact limitation we have currently. One workaround is to resolve the ALIASes on a hidden primary that signs. If you want to talk more, please [open a Q&A Discussion](https://github.com/PowerDNS/pdns/discussions) :)

**Comment by haslersn** [comment]
That answers the questions, but the documentation could be improved. Specifically [here](https://doc.powerdns.com/authoritative/guides/alias.html#alias-and-dnssec), where this limitation is not very obvious.

**Comment by klaus-nicat** [comment]
> Since we obviously want signed records, I enabled `outgoing-axfr-expand-alias=yes` on the primary (and re-enabled PRESIGNED on the secondary zone).

That's not so easy. See https://github.com/PowerDNS/pdns/pull/14594 which fixed the issue by not using some internal shortcuts. But the drawback is, that during the ALIAS lookup a backend is blocked. So, slow ALIAS resolving would slow down the authoritative server. If it would be easy, it would have been done already :)

**Comment by klaus-nicat** [comment]
> Here to report that `ALIAS` RR TYPE on wildcard name doesn't work (no `A/AAAA` are synthesized) in `4.9.3` autoritative. Didn't want to create a separate bug issue for it since it is mentioned in the description of this one already. Please do so if you feel it should be dealt separately.

I stumpled about this limitation as well today. But after some thinking, it is not a limitation. For wildcards (which are not at zone apex) a CNAME can be used, causing much less trouble. ALIAS is only needed for apex and then it never will be a wildcard.

**Comment by Habbie** [comment]
Unless you also want other content on the wildcard, with ALIAS handling A/AAAA and the other content statically being other things. But that doesn't feel that common - I hope.

**Comment by aoxborrow** [comment]
> Unless you also want other content on the wildcard, with ALIAS handling A/AAAA and the other content statically being other things. But that doesn't feel that common - I hope.

that's exactly what we need. many thousands of subdomains with ALIAS handling A/AAAA, each with specific TXT records, fully DNSSEC signed

**Comment by mdraaijer** [comment]
This is still causing practical problems for hosting providers and administrators today.

I recently encountered exactly this issue with a PowerDNS hosting provider: ALIAS works correctly, and DNSSEC works correctly with a static A record, but enabling DNSSEC with the ALIAS-generated A response results in no `RRSIG A`. Google Public DNS and Cloudflare consequently return `SERVFAIL`.

I opened #18040 with a detailed description of this use case and a possible implementation approach. It was pointed out there that this is a duplicate of this existing issue, specifically the still-open "support dynamic signing" item.

I would also like to emphasize why ALIAS is important in practice.

A hosting provider or administrator may have many domains hosted on the same server/IP address. Without ALIAS at the zone apex, every domain requires its own static A/AAAA record. If the server's IP address changes, all of those records have to be updated individually.

With ALIAS, all apex domains can instead point to one centrally managed hostname:

`domain1.example → ALIAS → server.example.net`
`domain2.example → ALIAS → server.example.net`
`domain3.example → ALIAS → server.example.net`
`...`

The server IP then only needs to be changed in one place. This becomes increasingly valuable when managing tens, hundreds, or potentially thousands of domains, and is particularly useful where server addresses are not guaranteed to remain static.

So this is not merely a convenience for avoiding a CNAME restriction at the zone apex. ALIAS provides an important layer of indirection that greatly simplifies centralized DNS administration and reduces the risk of configuration errors during IP address changes.

At present, administrators effectively have to choose between this centralized management benefit and DNSSEC when using ALIAS in normal authoritative operation.

Ten years after this issue was opened, support for dynamic signing of ALIAS records would therefore still provide significant practical value, especially for hosting providers managing large numbers of domains.
